Solution.cpp

```diff
@@ -220,6 +221,59 @@ void Solution::performHeuristic(){
 	}
 }
 
+void Solution::tabuSearch(double maxTime)
+{
+	int k = problem->n/10;
+	vector<TabuCandidate> candidates;
+	vector<vector<int>> tabuList( problem->n , vector<int> (problem->n, 0));
+	clock_t begin = clock();
+	int withoutImprovement=0;
+	do
+	{	
+		candidates.clear();
+		while (hasNextNeighbour())
+		{
+			tuple<int, int> swap = nextSwap;
+			int change = checkNextNeighbour();
+			if(tabuList[get<0>(swap)][get<1>(swap)]>0){
+				tabuList[get<0>(swap)][get<1>(swap)]--;
+				if(objectiveValue + change >= bestObjectiveValue) continue; //aspiration 
+			}
+			TabuCandidate candidate;
+			candidate.swap = swap;
+			candidate.change = change;
+			candidates.push_back(candidate);
+		}
+		sort( candidates.begin( ), candidates.end( ), [ ]( const auto& a, const auto& b ){return a.change < b.change;});
+		int size = min(k, (int)candidates.size());
+		candidates.resize(size);
+
+		int threshold = candidates[size-1].change;
+		int i = 0;
+		while(candidates[i].change<=threshold && i < candidates.size()){
+			//Make step
+			objectiveValue += candidates[i].change;
+			swap(permutation[get<0>(candidates[i].swap)], permutation[get<1>(candidates[i].swap)]);
+			if(objectiveValue < bestObjectiveValue){
+				withoutImprovement = 0;
+			}else{
+				withoutImprovement++;
+			}
+			updateBestSolution();
+			tabuList[get<0>(candidates[i].swap)][get<1>(candidates[i].swap)] = problem->n/4;
+			i++;
+
+			//Update changes for the rest
+			for(int j=i; j<candidates.size(); j++)
+			{
+				candidates[j].change = calcObjectValueChange(get<0>(candidates[j].swap),get<1>(candidates[j].swap));
+			}
+			sort( candidates.begin( ) + i, candidates.end( ), [ ]( const auto& a, const auto& b ){return a.change < b.change;});
+		}
+		nextSwap = make_tuple(1, 0);
+	} while (withoutImprovement < problem->n);
```

Comment:
Shouldn't it be dependent on size of neighborhood [for 2_OPT L = this->problem->n * (this->problem->n - 1) / 2] rather than instance size? We can multiply that L by some constant number like 10 according to the http://www.cs.put.poznan.pl/mkomosinski/site/?q=mioib-implementacja.